core/src/main/java/org/infinispan/encoding/DataConversion.java

```diff
@@ -91,26 +94,29 @@ private DataConversion(Encoder encoder, Wrapper wrapper, boolean isKey) {
       this.isKey = isKey;
       this.storageMediaType = MediaType.APPLICATION_OBJECT;
       this.requestMediaType = MediaType.APPLICATION_OBJECT;
+      encoderId = EncoderIds.NO_ENCODER;
+      wrapperId = WrapperIds.NO_WRAPPER;
    }
 
    public DataConversion withRequestMediaType(MediaType requestMediaType) {
       if (Objects.equals(this.requestMediaType, requestMediaType)) return this;
-      return new DataConversion(null, this.wrapperClass, requestMediaType, this.storageMediaType,
-            this.isKey);
+      return new DataConversion(null, this.wrapperClass, requestMediaType, this.isKey);
    }
 
    public DataConversion withEncoding(Class<? extends Encoder> encoderClass) {
       if (Objects.equals(this.encoderClass, encoderClass)) return this;
-      return new DataConversion(encoderClass, this.wrapperClass, this.requestMediaType, this.storageMediaType,
-            this.isKey);
+      return new DataConversion(encoderClass, this.wrapperClass, this.requestMediaType, this.isKey);
    }
 
    public DataConversion withWrapping(Class<? extends Wrapper> wrapperClass) {
       if (Objects.equals(this.wrapperClass, wrapperClass)) return this;
-      return new DataConversion(this.encoderClass, wrapperClass, this.requestMediaType, this.storageMediaType,
-            this.isKey);
+      return new DataConversion(this.encoderClass, wrapperClass, this.requestMediaType, this.isKey);
    }
 
+   /**
+    * @deprecated Since 11.0, will be removed with no replacement
+    */
+   @Deprecated
```

Comment:
I don't think we can deprecate this and remove with no replacement, the problem is RemoteQuery uses a complicated Wrapper that is not part of core, and it drags lots of dependencies to Hibernate Search.

Reply:
It's yet to see how the role this wrapper will play after the Hibernate Search 6 migration though...

Reply:
@gustavonalle is there a description somewhere of why we need `messageDescriptor` as a field in `ProtobufValueWrapper` instead of lazily computing it from the first few bytes of the payload when needed?

Reply:
How could it be computed? AFAICT, ```messageDescriptor``` is the schema that describe the blob

Reply:
@anistor can tell more about it

Reply:
The description is in the javadocs of ProtobufValueWrapper, I remember I wrote something there @danberindei . ProtobufValueWrapper is required for indexing. It needs that messageDescriptor to be able to decode the protobuf object. No, it cannot be really lazily computed without any external help.  ProtobufValueWrapper can obtain the type name or type id from the first few bytes but then it need a reference to a SerializationContext to be able to obtain that descriptor. So a bit more moving parts are involved here. What it the reason to remove/deprecate? And what ? :) [I did not follow discussion here and on zulip]

Reply:
@anistor deprecation would be a method in the cache that overrides a wrapper after the cache is created, and since ```ProtobufValueWrapper``` is not part of core, it'd be more difficult to associated it with a cache at creation time rather than in the RemoteQuery server lifecycle methods, but Dan has a proposal for it

Reply:
@gustavonalle I have updated the PR to just move `overrideWrapper()` from `DataConversion` to a component named `ConversionManager` for now. I got a bit carried away and made a lot more changes to the initialization of the cache wrappers, so the PR is much bigger now.

@anistor the javadoc in `ProtobufValueWrapper` says just

`The Descriptor of the message (if it's a Message and not a primitive value, or null otherwise). Transient field!`

AFAICT `WrappedMessageTagHandler` reads a `WRAPPED_DESCRIPTOR_FULL_NAME` field from the payload and computes the descriptor based on that, so I was thinking maybe `ProtobufValueWrapperFieldBridge` could also use a `WrappedMessageTagHandler` to compute a descriptor on demand.

Reply:
@danberindei I'm glad the very informative javadoc of ```ProtobufValueWrapper.messageDescriptor``` field piqued your interest. But you may want to read also the javadoc of the setter of this field or even follow the (unfortunately) half-broken link from that description, or alternatively read the code :).

messageDescriptor field is set by WrappedMessageTagHandler once, to (possibly) be accessed many times and avoid the penalty of being recomputed on demand many times. That field must exist to cache this expensive thing.

Two places where it is used are ProtobufValueWrapperAnalyzerDiscriminator and ProtobufValueWrapperIndexingInterceptor.

ProtobufValueWrapperFieldBridge is cache-scoped, not entry-scoped. So it must store that descriptor somewhereelse .... 

I do not understand yet what and why you need to change ProtobufValueWrapper and his many ugly friends @danberindei

Reply:
@anistor I do not intend to change `ProtobufValueWrapper`, but I am curious about why it is needed, so I asked "Is there a description somewhere of why we need messageDescriptor as a field in ProtobufValueWrapper instead of lazily computing it from the first few bytes of the payload when needed?"

So far I haven't seen an explanation in the javadoc or here of *why* computing the `messageDescriptor` lazily is so expensive that it's worth keeping the extra field in the heap at all times. I was imagining that the parsing of the message descriptor would be just reading a few bytes at a known offset in the value `byte[]`, is it much more than that?

Reply:
The primordial need for ProtobufValueWrapper is rooted in hibernate-search 5. It needs to discriminate on the class of an object (plus eventually configured class bridges) to be able to index it. But in 6 this changes a lot. I'd keep this ProtobufValueWrapper undisturbed until we have 6 merged, then we can try to re-analyse its worthiness.

Reply:
Ok, fair enough